### PR TITLE
[BACKPORT] Fix collapseContiguousMerges() being too  greedy with unit dimensions

### PR DESCRIPTION
**Pull request by krzysz00**
Fixes https://github.com/ROCm/rocMLIR/issues/1407

The bug resolved here is one that comes from an overly aggressive approach to adding length-1 dimensions to contiguous groups when detecting Merge/Unmerge pairs, which led to unusual permutation sequences like the one abstracted into the unit test for this commit causing an overeager dimension collapse. This led to runtime page faults because the H*W index into the output was being scaled by the large multiplier for the batch (N) dimension of the result tensor, leading to out of bounds accesses.

We guard against such issues in the future by strictly limiting the unit dimension grabber to only include those unit dimensions that lie within the contiguous group to be merged (as in the original Merge{8, 1, 3}, Unmerge{8. 3} example).

In addition, we add rudimentary debug output for merge collapsing so that it's clear that it's occurring.

### CODE REVIEW

File: mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp [hunk 1]
```diff
@@ -321,19 +321,30 @@ void findCountiguousGroupsUnmerge(const ArrayRef<uint32_t> upperDims,
...
+      // unsound to collapse onto trailing unit dimensions not otherwise
+      // accounted for above, as they could be permuted out to
+      // arbitrary other positions. Therefore, we only examine unit dimensions
+      // which come earlier in the merge parameters than the first true
+      // collapsed dimension.
+      ArrayRef<uint32_t> thisMergeDims = keyI.transform.getLowerDims();
+      ArrayRef<int64_t> thisMergeParams = keyI.transform.getParams();
+      for (const auto [d, p] : llvm::zip(
+               thisMergeDims.slice(slowestDimPosInMerge,
+                                   fastestDimPosInMerge - slowestDimPosInMerge),
```
Comment (sjw36) [inline mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:342]: Nit: for readability:
> int distance = fastestDimPosInMerge - slowestDimPosInMerge;
> thisMergDims.slice(slowestDimPosInMerge, distance);

**Review by sjw36 — APPROVED** [review]
Looks good.

File: mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp (hunk 1 above)
Comment (krzysz00) [reply mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp:342]: Think I should address that here? This is a backport PR, after all

**Review by manupak — APPROVED** [review]
Approving backport!

**Pull request merged**